### PR TITLE
CBG-1535 - Enforce TLS by default

### PR DESCRIPTION
**Pull request by IsaacLambat**
- [x] http://uberjenkins.sc.couchbase.com:8080/job/sync-gateway-integration/929/ (test TestDBOnlineWithDelayAndImmediate failed on integration but works locally)

https://issues.couchbase.com/browse/CM-837
Default bucket is no server is provided is now couchbases://localhost:8091 regardless of flag
If the server is a Walrus bucket, then this is treated as a secure server meaning UseTLSServer does not need to be set.

See https://docs.google.com/document/d/1Az-Fa75WebU1cokZQ1Xh6EOcL1JeqKcCqKso9rr6Jss/edit for config details

### CODE REVIEW

**Comment by bbrks** [comment]
Need to do the work to get the `sgw-pipeline-litecore-ee` stage passing (pass in a command line flag for this new option from the litecore test script)

I can help get you started with above when you have chance to look at this @IsaacLambat

File: rest/config_legacy.go [hunk 1]
```diff
@@ -117,12 +117,13 @@ func (lc *LegacyServerConfig) ToStartupConfig() (*StartupConfig, DbConfigMap, er
...
-			CACertPath:   dbConfig.CACertPath,
-			X509CertPath: dbConfig.CertPath,
-			X509KeyPath:  dbConfig.KeyPath,
+			Server:                         *dbConfig.Server,
+			Username:                       dbConfig.Username,
+			Password:                       dbConfig.Password,
+			CACertPath:                     dbConfig.CACertPath,
+			X509CertPath:                   dbConfig.CertPath,
+			X509KeyPath:                    dbConfig.KeyPath,
+			AllowInsecureServerConnections: base.BoolPtr(true),
```
Comment (JRascagneres) [inline rest/config_legacy.go]: Still need this to be false by default in the default case. Think this should mean just removing this as the current DefaultStartupConfig() should handle it

File: rest/config_legacy.go [hunk 2]
```diff
@@ -133,6 +134,9 @@ func (lc *LegacyServerConfig) ToStartupConfig() (*StartupConfig, DbConfigMap, er
 			Pretty:             lc.Pretty,
 			CompressResponses:  lc.CompressResponses,
 			HideProductVersion: lc.HideProductVersion,
+			HTTPS: HTTPSConfig{
+				AllowInsecureTLSConnections: base.BoolPtr(true),
```
Comment (JRascagneres) [inline rest/config_legacy.go]: Same as above

**Review by JRascagneres — CHANGES_REQUESTED** [review]

File: rest/config.go [hunk 3]
```diff
@@ -72,7 +72,7 @@ type BucketConfig struct {
 
 func (bc *BucketConfig) MakeBucketSpec() base.BucketSpec {
 
-	server := "http://localhost:8091"
+	server := "couchbases://localhost:8091"
```
Comment (bbrks) [inline rest/config.go]: 8091 is a http port, need to remove to allow it to default to 11207 ([memcached_ssl_port](https://docs.couchbase.com/server/current/install/install-ports.html#detailed-port-description))

```suggestion
	server := "couchbases://localhost"
```

File: rest/config.go [hunk 4]
```diff
@@ -809,6 +809,21 @@ func (config *StartupConfig) Serve(addr string, handler http.Handler) error {
 	)
 }
 
+// Validate insecure connections
+func (sc *StartupConfig) validateInsecureTLSConnection() (err error) {
+	// Validate SSL is provided if not allowing unsecure connections
+	if sc.API.HTTPS.AllowInsecureTLSConnections == nil || !*sc.API.HTTPS.AllowInsecureTLSConnections {
+		if sc.API.HTTPS.TLSKeyPath == "" || sc.API.HTTPS.TLSCertPath == "" {
+			return fmt.Errorf("a TLS key and cert path must be provided when not allowing insecure TLS connections")
```
Comment (bbrks) [inline rest/config.go]: The double negative in this message is a bit confusing. Think we can rephrase this to be:

```suggestion
			return fmt.Errorf("TLS key path and cert path must be provided unless api.https.allow_insecure_tls_connections is set")
```

File: rest/config.go [hunk 5]
```diff
@@ -809,6 +809,21 @@ func (config *StartupConfig) Serve(addr string, handler http.Handler) error {
...
+		}
+	} else { // Make sure TLS key and cert is not provided
+		if sc.API.HTTPS.TLSKeyPath != "" || sc.API.HTTPS.TLSCertPath != "" {
+			return fmt.Errorf("cannot use TLS and also use insecure TLS connections")
```
Comment (bbrks) [inline rest/config.go]: likewise

```suggestion
			return fmt.Errorf("cannot use TLS with allow_insecure_tls_connections set")
```

File: rest/config_legacy.go [hunk 6]
```diff
@@ -427,7 +427,7 @@ func ParseCommandLine(args []string, handling flag.ErrorHandling) (*LegacyServer
 	_ = flagSet.Bool("api.admin_interface_authentication", true, "")
 	_ = flagSet.Bool("api.metrics_interface_authentication", true, "")
 	_ = flagSet.Bool("bootstrap.allow_insecure_server_connections", false, "")
-	_ = flagSet.Bool("bootstrap.allow_insecure_tls_connections", false, "")
+	_ = flagSet.Bool("api.https.allow_insecure_tls_connections", false, "")
```
Comment (bbrks) [inline rest/config_legacy.go]: I think this is already done, which means the PR should be rebased onto the latest master.

File: rest/config_startup.go [hunk 7]
```diff
@@ -106,9 +109,10 @@ type APIConfig struct {
 }
 
 type HTTPSConfig struct {
-	TLSMinimumVersion string `json:"tls_minimum_version,omitempty"     help:"The minimum allowable TLS version for the REST APIs"`
-	TLSCertPath       string `json:"tls_cert_path,omitempty" help:"The TLS cert file to use for the REST APIs"`
-	TLSKeyPath        string `json:"tls_key_path,omitempty"  help:"The TLS key file to use for the REST APIs"`
+	TLSMinimumVersion           string `json:"tls_minimum_version,omitempty"     help:"The minimum allowable TLS version for the REST APIs"`
+	TLSCertPath                 string `json:"tls_cert_path,omitempty" help:"The TLS cert file to use for the REST APIs"`
+	TLSKeyPath                  string `json:"tls_key_path,omitempty"  help:"The TLS key file to use for the REST APIs"`
+	AllowInsecureTLSConnections *bool  `json:"allow_insecure_tls_connections" help:"Allow TLS to be optional for the REST APIs"`
```
Comment (bbrks) [inline rest/config_startup.go]: It's not just optional, since we fail to start up if you provide TLS cert/key paths.

My phrasing below still isn't perfect but is a bit clearer.

```suggestion
	AllowInsecureTLSConnections *bool  `json:"allow_insecure_tls_connections" help:"Allows REST APIs to start without TLS"`
```

File: rest/server_context.go [hunk 8]
```diff
@@ -372,6 +372,17 @@ func (sc *ServerContext) _getOrAddDatabaseFromConfig(config DatabaseConfig, useE
 		return nil, err
 	}
 
+	secure := spec.IsTLS()
+	if sc.config.Bootstrap.AllowInsecureServerConnections == nil || !*sc.config.Bootstrap.AllowInsecureServerConnections {
+		if !secure && !spec.IsWalrusBucket() {
+			return nil, fmt.Errorf("couchbase server URL must use secure protocol when disallowing insecure server connections. Current URL: %s", spec.Server)
```
Comment (bbrks) [inline rest/server_context.go]: "disallowing insecure connections" is the default here, so again the double negative is just going to cause confusion for users encountering this (they won't know about that option).


```suggestion
			return nil, fmt.Errorf("Must use secure scheme in Couchbase Server URL, or opt out using allow_insecure_server_connections: %s", spec.Server)
```

File: rest/utilities_testing.go [hunk 9]
```diff
@@ -42,17 +42,19 @@ import (
...
+	DatabaseConfig                *DbConfig                  // Supports additional config options.  BucketConfig, Name, Sync, Unsupported will be ignored (overridden)
+	InitSyncSeq                   uint64                     // If specified, initializes _sync:seq on bucket creation.  Not supported when running against walrus
+	EnableNoConflictsMode         bool                       // Enable no-conflicts mode.  By default, conflicts will be allowed, which is the default behavior
+	distributedIndex              bool                       // Test with walrus-based index bucket
+	TestBucket                    *base.TestBucket           // If set, use this bucket instead of requesting a new one.
+	adminInterface                string                     // adminInterface overrides the default admin interface.
+	sgReplicateEnabled            bool                       // sgReplicateManager disabled by default for RestTester
+	sgr1Replications              []*ReplicateV1ConfigLegacy // sgr1Replications are a list of replications to enable on the server context.
+	hideProductInfo               bool
+	denyInsecureTLSConnections    bool // If true, TLS will be used with SG
```
Comment (bbrks) [inline rest/utilities_testing.go]: Is it `denyInsecure...` here rather than `allowInsecure...` so that the zero value of false matches SG's default of true?

Is Could that be a bit misleading when testing?

**Review by bbrks — CHANGES_REQUESTED** [review]
I think it looks OK besides rephrasing some of the error messages to make them more user-friendly and some small cleanup

File: rest/admin_api_test.go [hunk 10]
```diff
@@ -80,6 +78,8 @@ func TestPutDocSpecialChar(t *testing.T) {
 	}
 	for _, testCase := range testCases {
 		t.Run(testCase.name, func(t *testing.T) {
+			rt := NewRestTester(t, nil)
```
Comment (JRascagneres) [inline rest/admin_api_test.go]: Any reason we've moved this? Seems unrelated to this PR

File: rest/admin_api_test.go [hunk 11]
```diff
@@ -92,6 +92,8 @@ func TestPutDocSpecialChar(t *testing.T) {
 	}
 
 	t.Run("Delete Double quote Doc ID", func(t *testing.T) { // Should be done for Local Document deletion when it returns response
+		rt := NewRestTester(t, nil)
```
Comment (JRascagneres) [inline rest/admin_api_test.go]: ditto

File: rest/config_legacy.go [hunk 12]
```diff
@@ -112,6 +115,9 @@ func (lc *LegacyServerConfig) ToStartupConfig() (*StartupConfig, DbConfigMap, er
 			Pretty:             lc.Pretty,
 			CompressResponses:  lc.CompressResponses,
 			HideProductVersion: lc.HideProductVersion,
+			HTTPS: HTTPSConfig{
+				UseTLSClient: lc.UseTLSClient,
```
Comment (JRascagneres) [inline rest/config_legacy.go:127]: What is this going to end up being if its not specified in the config? Defaults to false maybe before flag / Default overrides?

**Review by JRascagneres — CHANGES_REQUESTED** [review]
Just stuck a few comments on and discussion in person about other bits, mostly around defaults / flags

File: rest/config_legacy.go (hunk 12 above)
Comment (IsaacLambat) [reply rest/config_legacy.go:127]: Tested it and the default of true does get applied if the user has not specified it in their config.

**Review by bbrks — CHANGES_REQUESTED** [review]
help text needs updating - and need to investigate why RestTester works with the client use TLS option defaulting to true (shouldn't it fail to start up the server because we don't have any HTTPS certs?)

**Review by bbrks — APPROVED** [review]

**Review by JRascagneres — APPROVED** [review]
